torch/_inductor/scheduler.py

```diff
@@ -861,6 +886,189 @@ def check_all_pairs(nodes):
+                ):
+                    continue
+
+                def node_ndim(node):
+                    pointwise, reduction = node.get_ranges()
+                    return len(pointwise) + len(reduction)
+
+                ndim = max(node_ndim(n) for n in node1.get_nodes())
+
+                def filter_in(dep):
+                    if not isinstance(dep, MemoryDep):
+                        return False
+                    return (
+                        isinstance(dep, MemoryDep)
+                        and dep.name == buf_name
+                        and len(dep.size) == ndim
+                    )
+
+                node1_buf_deps = [
+                    dep
+                    for dep in (node1.read_writes.reads | node1.read_writes.writes)
+                    if filter_in(dep)
+                ]
+                node2_buf_deps = [
+                    dep
+                    for dep in (node2.read_writes.reads | node2.read_writes.writes)
+                    if filter_in(dep)
+                ]
+
+                for dep1, dep2 in itertools.product(node1_buf_deps, node2_buf_deps):
+                    if dep1 == dep2:
+                        continue
+
+                    reorder = self.reorder_deps(dep1, dep2)
+                    if reorder is not None:
+                        reorder_opportunities[(node1, node2)].add(tuple(reorder))
+
+        def reorder_loop_node(
+            node: SchedulerNode, new_order: Tuple[int, ...]
+        ) -> SchedulerNode:
+            reorder = ir.same_reorder(order)
+
+            def loop_reorder(ir_node: ir.ComputedBuffer):
+                (
+                    iter_vars,
+                    reduce_vars,
+                ), var_ranges = dependencies.index_vars_no_squeeze(
+                    *node._sizes, prefix="sched_reorder"
+                )
+                assert len(reduce_vars) == 0
+                body = ir.LoopBody(
+                    node._body,
+                    [reorder(iter_vars), reduce_vars],
+                    var_ranges,
+                )
+                iter_ranges = [var_ranges[v] for v in reorder(iter_vars)]
+                reduce_ranges = []
+                return (iter_ranges, reduce_ranges), body
+
+            new_node = SchedulerNode(
+                self,
+                node.node,
+                self.get_backend(node.get_device()).group_fn,
+                loop_reorder=loop_reorder,
+            )
+            new_node.users = node.users
+            new_node.inverse_users = node.inverse_users
+            new_node.recursive_predecessors = node.recursive_predecessors
+            new_node.min_order = node.min_order
+            new_node.max_order = node.max_order
+            new_node.last_usage = node.last_usage
+            new_node.written = node.written
+            return new_node
+
+        for (node1, node2), orders in reorder_opportunities.items():
+            for order in orders:
```

Comment:
I changed some the names to hopefully clarify whats going on.

`orders` here is the set of all permutations returned from `pick_dependency_reorder`, but at this stage we haven't actually reordered anything.

For each permutation we then call `node.reorder_loop`  and the dependency information for the new loop is calculated in `SchedulerNode`'s constructor.

`can_fuse` then does all the usual checks, including checking that memory dependencies match which in theory we know should at least match for the dependency that `pick_dependency_reorder` returned.

We could maybe do more to prune `orders` down before trying them for real, but I'm not sure it would be worth the added code complexity.